### PR TITLE
RandIntLayer: allow shape from layer

### PR DESCRIPTION
**Pull request by vieting**
For https://github.com/rwth-i6/pytorch-to-returnn/pull/98, we need the `RandIntLayer` to allow shapes based on other layers. In the added test case, we e.g. create rand ints with the shape `(B, T)`. The question is how to obtain the `dim_tag` if the input layer is not a `LengthLayer`, e.g. for the time dim in the test cast as the output of the `LengthLayer` is reduced here.

### CODE REVIEW

File: returnn/tf/layers/basic.py [hunk 1]
```diff
@@ -2290,16 +2293,17 @@ def transform_config_dict(cls, d, network, get_layer):
     """
     d.setdefault("from", [])
     super(RandIntLayer, cls).transform_config_dict(d, network=network, get_layer=get_layer)
-    if isinstance(d.get("maxval", None), str):
-      d["maxval"] = get_layer(d["maxval"])
-    if isinstance(d.get("minval", None), str):
-      d["minval"] = get_layer(d["minval"])
+
+    d["maxval"] = get_layer(d["maxval"]) if isinstance(d["maxval"], str) else d["maxval"]
+    d["minval"] = get_layer(d["minval"]) if isinstance(d["minval"], str) else d["minval"]
```
Comment (albertz) [inline returnn/tf/layers/basic.py:2298]: Why this change for maxval/minval?

File: returnn/tf/layers/basic.py [hunk 2]
```diff
@@ -2290,16 +2293,17 @@ def transform_config_dict(cls, d, network, get_layer):
...
+    d["shape"] = tuple(get_layer(dim) if isinstance(dim, str) else dim for dim in d["shape"])
+
```
Comment (albertz) [inline returnn/tf/layers/basic.py:2300]: One empty line too much.

**Comment by albertz** [comment]
> The question is how to obtain the `dim_tag` if the input layer is not a `LengthLayer`, e.g. for the time dim in the test cast as the output of the `LengthLayer` is reduced here.

Not all layer outputs correspond to an existing dim tag.

Via `Dim.get_tag_from_size_tensor` on the `layer.output.placeholder`, you have a generic way to find any corresponding existing dim tag.

If it does not exist, you could also just create a new dim tag.

I think we should not try to be too clever on RETURNN side on this. I think we can much easier handle this on PyTorch-to-RETURNN side. I actually wonder, why do you need to pass layers at all? You can simply pass dim tags directly (which you get from `SizeValue`). What's the problem with that?

**Comment by albertz** [comment]
Btw, somewhat related might be #920 and #918.

**Comment by vieting** [comment]
> Via `Dim.get_tag_from_size_tensor` on the `layer.output.placeholder`, you have a generic way to find any corresponding existing dim tag.

Thanks for the hint. In the test case here, this returns `None` in both cases (even for the `LengthsLayer`)

> I think we should not try to be too clever on RETURNN side on this. I think we can much easier handle this on PyTorch-to-RETURNN side. I actually wonder, why do you need to pass layers at all? You can simply pass dim tags directly (which you get from SizeValue). What's the problem with that?

Currently we always do `x = x.as_tensor()` if `isinstance(x, SizeValue) and x.dim_tag.dimension is None` [here](https://github.com/rwth-i6/pytorch-to-returnn/blob/878f62db88bbb0ed2c362d38f834082f4454dae7/pytorch_to_returnn/naming/naming.py#L205) in `self._make_tensor`. We could redefine this logic to use the dim tags directly for `RandInt`. Would that work also for the printed network dict?

**Comment by albertz** [comment]
> > Via `Dim.get_tag_from_size_tensor` on the `layer.output.placeholder`, you have a generic way to find any corresponding existing dim tag.
> 
> Thanks for the hint. In the test case here, this returns `None` in both cases (even for the `LengthsLayer`)

This should not be. This might be another bug then.

> > I think we should not try to be too clever on RETURNN side on this. I think we can much easier handle this on PyTorch-to-RETURNN side. I actually wonder, why do you need to pass layers at all? You can simply pass dim tags directly (which you get from SizeValue). What's the problem with that?
> 
> Currently we always do `x = x.as_tensor()` if `isinstance(x, SizeValue) and x.dim_tag.dimension is None` [here](https://github.com/rwth-i6/pytorch-to-returnn/blob/878f62db88bbb0ed2c362d38f834082f4454dae7/pytorch_to_returnn/naming/naming.py#L205) in `self._make_tensor`. We could redefine this logic to use the dim tags directly for `RandInt`.

Ah yes. No, this should not be changed. But you can get back the dim tag via `tensor_entry.is_dim`.

> Would that work also for the printed network dict?

We anyway make use of dim tags in other parts, and then also here. For the serialization of dim tags, we already spoke about that, this can be done. We did not implement it yet because it was not needed for the tests. Once we need that, we can follow a very similar implementation as in returnn-common.

One other small problem we might get here: Via dim tags, we will not properly derive dependencies. E.g. when the dim tag is some dynamic length originating from some other layer, when this dim tag is just used elsewhere, it might not be defined yet. To make sure it gets defined, we however can add it as explicit dependency. E.g. the `RandIntLayer` could just use `from`. It will ignore `from` otherwise. See here for example: https://github.com/rwth-i6/returnn/blob/0b49ab4c95911c040453e00f8358805e7560096a/tests/test_TFNetworkLayer.py#L6510

**Comment by vieting** [comment]
> One other small problem we might get here: Via dim tags, we will not properly derive dependencies. E.g. when the dim tag is some dynamic length originating from some other layer, when this dim tag is just used elsewhere, it might not be defined yet. To make sure it gets defined, we however can add it as explicit dependency. E.g. the RandIntLayer could just use from. It will ignore from otherwise. See here for example: 

Yes I noticed this problem already. Is there a general answer to what should be the layer we use as "from" here?

**Comment by albertz** [comment]
> > One other small problem we might get here: Via dim tags, we will not properly derive dependencies. E.g. when the dim tag is some dynamic length originating from some other layer, when this dim tag is just used elsewhere, it might not be defined yet. To make sure it gets defined, we however can add it as explicit dependency. E.g. the RandIntLayer could just use from. It will ignore from otherwise. See here for example:
> 
> Yes I noticed this problem already. Is there a general answer to what should be the layer we use as "from" here?

The `CopyLayer` has `extra_deps`. So you could add a `copy` on any of the inputs.

This would only work for layers with inputs however but not e.g. for `RandIntLayer`. For `RandIntLayer` however, we could use `from`. But still there are maybe others, e.g. `ConstantLayer` or so.

Maybe we should just add `extra_deps` to `LayerBase`. There was some technical reason why I did not do this but I don't exactly remember.

**Comment by vieting** [comment]
My question was regarding what layer I should use here for `RandIntLayer` in `from` or if there is some specific layer that should be added and used as `from`.

**Comment by albertz** [comment]
> My question was regarding what layer I should use here for `RandIntLayer` in `from` or if there is some specific layer that should be added and used as `from`.

The `SizeValue.originating_tensor` should be the right one.

**Comment by albertz** [comment]
I guess this PR can be closed, right? From the discussion, I think we agreed that we should just use dim tags instead.

**Comment by vieting** [comment]
> I guess this PR can be closed, right? From the discussion, I think we agreed that we should just use dim tags instead.

Yes, I agree

> > > Via `Dim.get_tag_from_size_tensor` on the `layer.output.placeholder`, you have a generic way to find any corresponding existing dim tag.
> > 
> > 
> > Thanks for the hint. In the test case here, this returns `None` in both cases (even for the `LengthsLayer`)
> 
> This should not be. This might be another bug then.

I could add an issue for this if you like.

**Comment by albertz** [comment]
> > > > Via `Dim.get_tag_from_size_tensor` on the `layer.output.placeholder`, you have a generic way to find any corresponding existing dim tag.
> > > 
> > > Thanks for the hint. In the test case here, this returns `None` in both cases (even for the `LengthsLayer`)
> > 
> > This should not be. This might be another bug then.
> 
> I could add an issue for this if you like.

I think not so important, unless we run in it at some point. In general I think we should not use `Dim.get_tag_from_size_tensor` too much.
